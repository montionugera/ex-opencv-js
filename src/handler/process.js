import { Errors } from 'moleculer'
import { generateFields } from 'services/vision'
import i18n, { initI18n } from 'lib/i18n'

import process from 'process'
import preprocess from 'preprocess'
import postprocess from 'postprocess'
import formatter from 'formatter'

export default {
  name: 'ocr',
  async started() {
    await initI18n()
  },
  actions: {
    process: {
      bulkhead: {
        enabled: true,
        concurrency: 5,
        maxQueueSize: Infinity,
      },
      async handler(ctx) {
        try {
          const { files, type } = ctx.params
          const { buffer } = files[0]

          const tracing = this.broker.createTracing(ctx)

          // tracing.start('preprocess')
          // const base64 = await preprocess(Buffer.from(buffer))
          // tracing.end('preprocess')

          const base64 = Buffer.from(buffer).toString('base64')

          tracing.start('vision')
          const scene = await generateFields(base64)
          tracing.end('vision')

          tracing.start('process')
          const result = await process(scene, type)
          tracing.end('process')

          tracing.start('postprocess')
          const postResult = await postprocess(
            Buffer.from(buffer),
            result,
            type
          )
          tracing.end('postprocess')

          const formatedResult = formatter(postResult)

          return {result : formatedResult, version : 1.0}
        } catch (err) {
          console.error(err)
          throw new Errors.MoleculerError(
            i18n().t('attachment:ocr.error.AIS.0101')
          )
        }
      },
    },
  },
}
