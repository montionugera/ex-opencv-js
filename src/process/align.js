/* eslint-disable no-bitwise */
/* eslint-disable new-cap */
import * as fp from 'lodash/fp'
import jsfeat from 'jsfeat'
import { updatePolygon } from 'selector'
import { isBetween } from 'utils'

export default (sceneList, { src, to, box }) => {
  const kernel = new jsfeat.motion_model.homography2d()
  const transformMatrix = new jsfeat.matrix_t(3, 3, jsfeat.F32_t | jsfeat.C1_t)
  const count = fp.size(src)

  // each point will be marked as good(1) or bad(0)
  const maskMatrix = new jsfeat.matrix_t(count, 1, jsfeat.U8_t | jsfeat.C1_t)

  const threshold = 3 // max error to classify as inlier
  const eps = 0.5 // max outliers ratio
  const prob = 0.99 // probability of success
  const params = new jsfeat.ransac_params_t(count, threshold, eps, prob)

  const maxIters = count

  // TODO - reject if isAlign is false
  const isAlign = jsfeat.motion_estimator.ransac(
    params,
    kernel,
    src,
    to,
    count,
    transformMatrix,
    maskMatrix,
    maxIters
  )

  let minX
  let minY
  let maxX
  let maxY

  if (box) {
    const xList = fp.pipe(fp.flatten, fp.map(fp.get('x')))(box)
    const yList = fp.pipe(fp.flatten, fp.map(fp.get('y')))(box)

    minX = fp.min(xList)
    minY = fp.min(yList)
    maxX = fp.max(xList)
    maxY = fp.max(yList)
  }

  const hasBox =
    fp.isFinite(minX) &&
    fp.isFinite(minY) &&
    fp.isFinite(maxX) &&
    fp.isFinite(maxY)

  return fp.pipe(
    updatePolygon(
      fp.ifElse(
        fp.any(({ x, y }) => {
          if (hasBox) {
            return isBetween(x, minX, maxX) && isBetween(y, minY, maxY)
          }

          return true
        }),
        fp.map(({ x, y }) => {
          const H = transformMatrix.data
          let w = H[6] * x + H[7] * y + H[8] * 1

          if (w === 0) {
            w = 1
          }

          return {
            x: Math.abs(fp.round((H[0] * x + H[1] * y + H[2] * 1) / w)),
            y: Math.abs(fp.round((H[3] * x + H[4] * y + H[5] * 1) / w)),
          }
        }),
        fp.constant([])
      )
    )
  )(sceneList)
}
