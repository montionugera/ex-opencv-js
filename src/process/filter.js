import * as fp from 'lodash/fp'
import { getText, getPolygon } from 'selector'

const filterByWord = (wordList, list) =>
  fp.pipe(
    fp.map((word) =>
      fp.pipe(
        fp.findIndex(
          fp.pipe(
            getText,
            fp.allPass([fp.pipe(fp.size, fp.lt(3)), fp.isEqual(word)])
            //
          )
        ),
        (index) => {
          if (index >= 0) {
            const [result] = list.splice(index, 1)

            return result
          }
        }
      )(list)
    ),
    fp.reject(fp.isNil)
  )(wordList)

const toPolygon = fp.pipe(
  fp.orderBy(['description'], ['asc']),
  fp.map(getPolygon),
  fp.flatten
  //
)

export default (baseList, sceneList, wordList) => {
  let filteredBaseList = fp.cloneDeep(baseList)
  let filteredSceneList = fp.cloneDeep(sceneList)

  if (wordList) {
    filteredBaseList = filterByWord(wordList, baseList)
    filteredSceneList = filterByWord(wordList, sceneList)
  }

  if (fp.size(filteredBaseList) < fp.size(filteredSceneList)) {
    const wordList = fp.map(getText, filteredBaseList)
    filteredSceneList = filterByWord(wordList, filteredSceneList)
  }

  if (fp.size(filteredSceneList) < fp.size(filteredBaseList)) {
    const wordList = fp.map(getText, filteredSceneList)
    filteredBaseList = filterByWord(wordList, filteredBaseList)
  }

  return {
    filteredBaseList: toPolygon(filteredBaseList),
    filteredSceneList: toPolygon(filteredSceneList),
  }
}
