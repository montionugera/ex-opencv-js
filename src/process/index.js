import {
  getPolygon,
  createPolygon,
  getTextAnnotation,
  getFullTextAnnotation,
  getSize,
} from 'selector'
import * as model from 'model'

import filter from './filter'
import align from './align'
import map from './map'

export default async (scene, type = 'car_registration_upper') => {
  const base = model.base[type]
  const { fields, whitelist } = model.mapping[type]

  const baseList = getTextAnnotation(base)
  const sceneForBaseList = getTextAnnotation(scene)
  const sceneList = getFullTextAnnotation(scene)

  const borderPolygon = filter(baseList, sceneForBaseList)
  const fieldsPolygon = filter(baseList, sceneForBaseList, whitelist)

  const { width, height } = getSize(base)

  const borderList = [
    createPolygon([
      { x: 0, y: 0 },
      { x: width, y: 0 },
      { x: width, y: height },
      { x: 0, y: height },
    ]),
  ]
  const [alignedBorderList] = align(borderList, {
    src: borderPolygon.filteredBaseList,
    to: borderPolygon.filteredSceneList,
  })

  const alignedSceneList = align(sceneList, {
    src: fieldsPolygon.filteredSceneList,
    to: fieldsPolygon.filteredBaseList,
    box: getPolygon(alignedBorderList),
  })

  const filteredSceneList = map(alignedSceneList, whitelist, fields)

  return filteredSceneList
}
