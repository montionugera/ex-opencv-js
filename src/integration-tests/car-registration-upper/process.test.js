import * as fp from 'lodash/fp'
import fs from 'fs'
import path from 'path'
import glob from 'glob'

import process from 'process'

const filterList = []

describe('process', () => {
  const files = glob.sync(path.join(__dirname, '../../__fixtures__/car-registration-upper/*.json'))

  const testCases = fp.pipe(
    fp.filter((path) => {
      if (fp.isEmpty(filterList)) return true

      const name = fp.pipe(
        fp.split('/'),
        fp.last
        //
      )(path)

      return fp.contains(name, filterList)
    }),
    fp.map((path) => {
      const name = fp.pipe(
        fp.split('/'),
        fp.last
        //
      )(path)
      const data = JSON.parse(fs.readFileSync(path))

      return {
        name,
        data,
      }
    })
  )(files)

  testCases
    .forEach(({ name, data }) => {
      describe(name, () => {
        it('return correctly', () => {
          const result = process(data)
          expect(result).toMatchSnapshot()
        })
      })
    })
})
