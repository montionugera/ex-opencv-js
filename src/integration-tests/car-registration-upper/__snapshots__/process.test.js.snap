// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process 1.json return correctly 1`] = `
Object {
  "confidence": Object {
    "vehicle_body_type": 0.58,
    "vehicle_brand": 0.96,
    "vehicle_car_weight": 0.77,
    "vehicle_chassis_location": 0.89,
    "vehicle_chassis_number": 0.69,
    "vehicle_color": 0.8,
    "vehicle_displacement": 0.96,
    "vehicle_engine_brand": 0.99,
    "vehicle_engine_location": 0.67,
    "vehicle_engine_number": 0.93,
    "vehicle_fuel_type": 0.94,
    "vehicle_gas_number": 0,
    "vehicle_gvw": 0.85,
    "vehicle_hp": 0.98,
    "vehicle_license_number": 0.73,
    "vehicle_license_province": 0.87,
    "vehicle_model": 0.93,
    "vehicle_model_year": 0.99,
    "vehicle_piston": 1,
    "vehicle_register_date": 0.92,
    "vehicle_seat_number": 0.93,
    "vehicle_shaft": 0.91,
    "vehicle_shaft_weight": 0,
    "vehicle_short_type": 0.69,
    "vehicle_type": 0.94,
  },
  "vehicle_body_type": "นั่งสามตอน",
  "vehicle_brand": "MITSUBISHI",
  "vehicle_car_weight": "1000",
  "vehicle_chassis_location": "หลังขวา",
  "vehicle_chassis_number": "MITGUKR1OHH005341",
  "vehicle_color": "เทา",
  "vehicle_displacement": "2442",
  "vehicle_engine_brand": "MITSUBISHI",
  "vehicle_engine_location": "ขวาเครื่อง",
  "vehicle_engine_number": "4N15UBR8093",
  "vehicle_fuel_type": "ดีเซล",
  "vehicle_gas_number": "",
  "vehicle_gvw": "1900",
  "vehicle_hp": "181",
  "vehicle_license_number": "617",
  "vehicle_license_province": "กรุงเทพมหานคร",
  "vehicle_model": "PAJERO SPORT",
  "vehicle_model_year": "2017",
  "vehicle_piston": "4",
  "vehicle_register_date": "26 มกราคม 2561",
  "vehicle_seat_number": "7",
  "vehicle_shaft": "2 เพลา 4 ล้อ ยาง 4 เส้น",
  "vehicle_shaft_weight": "",
  "vehicle_short_type": "“1",
  "vehicle_type": "รถยนต์นั่งส่วนบุคคลไม่เกิน 7 คน",
}
`;

exports[`process 2.json return correctly 1`] = `
Object {
  "confidence": Object {
    "vehicle_body_type": 0.58,
    "vehicle_brand": 0.95,
    "vehicle_car_weight": 0.76,
    "vehicle_chassis_location": 0.74,
    "vehicle_chassis_number": 0.79,
    "vehicle_color": 0.89,
    "vehicle_displacement": 0.97,
    "vehicle_engine_brand": 0.99,
    "vehicle_engine_location": 0.63,
    "vehicle_engine_number": 0.91,
    "vehicle_fuel_type": 0.62,
    "vehicle_gas_number": 0,
    "vehicle_gvw": 0.81,
    "vehicle_hp": 0.98,
    "vehicle_license_number": 0.65,
    "vehicle_license_province": 0.89,
    "vehicle_model": 0.96,
    "vehicle_model_year": 0.96,
    "vehicle_piston": 0.99,
    "vehicle_register_date": 0.97,
    "vehicle_seat_number": 0.92,
    "vehicle_shaft": 0.93,
    "vehicle_shaft_weight": 0,
    "vehicle_short_type": 0.45,
    "vehicle_type": 0.92,
  },
  "vehicle_body_type": "นั่งสามตอน",
  "vehicle_brand": "MITSUBISHI",
  "vehicle_car_weight": "1000",
  "vehicle_chassis_location": "หลังขวา",
  "vehicle_chassis_number": "HMIGUKR1oHH005341",
  "vehicle_color": "เทา",
  "vehicle_displacement": "2442",
  "vehicle_engine_brand": "MITSUBISHI",
  "vehicle_engine_location": "ขวาเครื่อง",
  "vehicle_engine_number": "4N15UBR8093",
  "vehicle_fuel_type": "ดีเซล -",
  "vehicle_gas_number": "",
  "vehicle_gvw": "1900",
  "vehicle_hp": "181",
  "vehicle_license_number": "6ช 617",
  "vehicle_license_province": "กรุงเทพมหานคร",
  "vehicle_model": "PAJERO SPORT",
  "vehicle_model_year": "2017",
  "vehicle_piston": "4",
  "vehicle_register_date": "26 มกราคม 2561",
  "vehicle_seat_number": "7",
  "vehicle_shaft": "2 เพลา 4 ล้อ ยาง 4 เส้น",
  "vehicle_shaft_weight": "",
  "vehicle_short_type": "51",
  "vehicle_type": "รถยนต์นั่งส่วนบุคคลไม่เกิน 7 คน",
}
`;

exports[`process 3.json return correctly 1`] = `
Object {
  "confidence": Object {
    "vehicle_body_type": 0.78,
    "vehicle_brand": 0.77,
    "vehicle_car_weight": 0,
    "vehicle_chassis_location": 0.67,
    "vehicle_chassis_number": 0.62,
    "vehicle_color": 0,
    "vehicle_displacement": 0.72,
    "vehicle_engine_brand": 0.55,
    "vehicle_engine_location": 0.8,
    "vehicle_engine_number": 0.82,
    "vehicle_fuel_type": 0.75,
    "vehicle_gas_number": 0,
    "vehicle_gvw": 0,
    "vehicle_hp": 0,
    "vehicle_license_number": 0.71,
    "vehicle_license_province": 0.9,
    "vehicle_model": 0.94,
    "vehicle_model_year": 0.97,
    "vehicle_piston": 0.99,
    "vehicle_register_date": 0.96,
    "vehicle_seat_number": 0,
    "vehicle_shaft": 0.76,
    "vehicle_shaft_weight": 0,
    "vehicle_short_type": 0.75,
    "vehicle_type": 0.57,
  },
  "vehicle_body_type": "รถจักรยานยนต์",
  "vehicle_brand": "HONDA",
  "vehicle_car_weight": "",
  "vehicle_chassis_location": "อยู่ที่ใต้ที่นั่ง",
  "vehicle_chassis_number": "ALIJF350XKS629846",
  "vehicle_color": "",
  "vehicle_displacement": "124.88",
  "vehicle_engine_brand": ": HONDA",
  "vehicle_engine_location": "ช้ายเครื่อง",
  "vehicle_engine_number": "JF35E-0229846",
  "vehicle_fuel_type": "เบนซิน",
  "vehicle_gas_number": "",
  "vehicle_gvw": "",
  "vehicle_hp": "",
  "vehicle_license_number": "9กร 7876",
  "vehicle_license_province": "กรุงเทพมหานคร",
  "vehicle_model": "CLICK 125 I",
  "vehicle_model_year": "2019",
  "vehicle_piston": "1",
  "vehicle_register_date": "11 มิถุนายน 2562",
  "vehicle_seat_number": "",
  "vehicle_shaft": "- เพลา 2 ล้อ ยาง 2 เส้น",
  "vehicle_shaft_weight": "",
  "vehicle_short_type": "12",
  "vehicle_type": "รถจักรยานยนต์2",
}
`;

exports[`process 4.json return correctly 1`] = `
Object {
  "confidence": Object {
    "vehicle_body_type": 0.72,
    "vehicle_brand": 0.65,
    "vehicle_car_weight": 0,
    "vehicle_chassis_location": 0.86,
    "vehicle_chassis_number": 0.94,
    "vehicle_color": 0.94,
    "vehicle_displacement": 0.9,
    "vehicle_engine_brand": 0.99,
    "vehicle_engine_location": 0.83,
    "vehicle_engine_number": 0.95,
    "vehicle_fuel_type": 0.77,
    "vehicle_gas_number": 0,
    "vehicle_gvw": 0,
    "vehicle_hp": 0,
    "vehicle_license_number": 0.81,
    "vehicle_license_province": 0.9,
    "vehicle_model": 0.97,
    "vehicle_model_year": 0.99,
    "vehicle_piston": 1,
    "vehicle_register_date": 0.96,
    "vehicle_seat_number": 0.99,
    "vehicle_shaft": 0.98,
    "vehicle_shaft_weight": 0,
    "vehicle_short_type": 0.99,
    "vehicle_type": 0.98,
  },
  "vehicle_body_type": "านยนต์",
  "vehicle_brand": "HONDA ,",
  "vehicle_car_weight": "",
  "vehicle_chassis_location": "ใต้ที่นั่ง",
  "vehicle_chassis_number": "HLHJF350XIK5629846",
  "vehicle_color": "ดำ",
  "vehicle_displacement": "124.88",
  "vehicle_engine_brand": "HONDA",
  "vehicle_engine_location": "ช้ายเครื่อง",
  "vehicle_engine_number": "JF35E-0229846",
  "vehicle_fuel_type": "เขนชิน",
  "vehicle_gas_number": "",
  "vehicle_gvw": "",
  "vehicle_hp": "",
  "vehicle_license_number": "9กร 7876",
  "vehicle_license_province": "กรุงเทพมหานคร",
  "vehicle_model": "CLICK 125 I",
  "vehicle_model_year": "2019",
  "vehicle_piston": "1",
  "vehicle_register_date": "11 มิถุนายน 2562",
  "vehicle_seat_number": "2",
  "vehicle_shaft": "- เพลา 2 ล้อ ยาง 2 เส้น",
  "vehicle_shaft_weight": "",
  "vehicle_short_type": "12",
  "vehicle_type": "รถจักรยานยนต์",
}
`;

exports[`process 5.json return correctly 1`] = `
Object {
  "confidence": Object {
    "vehicle_body_type": 0.45,
    "vehicle_brand": 0.95,
    "vehicle_car_weight": 0.83,
    "vehicle_chassis_location": 0.73,
    "vehicle_chassis_number": 0.71,
    "vehicle_color": 0.98,
    "vehicle_displacement": 0.97,
    "vehicle_engine_brand": 0.96,
    "vehicle_engine_location": 0.45,
    "vehicle_engine_number": 0.69,
    "vehicle_fuel_type": 0,
    "vehicle_gas_number": 0,
    "vehicle_gvw": 0.93,
    "vehicle_hp": 0.97,
    "vehicle_license_number": 0.84,
    "vehicle_license_province": 0.86,
    "vehicle_model": 0.86,
    "vehicle_model_year": 0.99,
    "vehicle_piston": 0.98,
    "vehicle_register_date": 0.87,
    "vehicle_seat_number": 0.63,
    "vehicle_shaft": 0.98,
    "vehicle_shaft_weight": 0,
    "vehicle_short_type": 0.41,
    "vehicle_type": 0.89,
  },
  "vehicle_body_type": "ฝั่งสามตอน",
  "vehicle_brand": "MITSUBISHI",
  "vehicle_car_weight": "1900",
  "vehicle_chassis_location": "หลังขวา",
  "vehicle_chassis_number": "แNTUKR1on1005341.",
  "vehicle_color": "เทา",
  "vehicle_displacement": "2442",
  "vehicle_engine_brand": "MITSUBISHI",
  "vehicle_engine_location": "ชวาเหวี่ยง",
  "vehicle_engine_number": "4Ni5UER8093",
  "vehicle_fuel_type": "",
  "vehicle_gas_number": "",
  "vehicle_gvw": "1900",
  "vehicle_hp": "181",
  "vehicle_license_number": "6 617",
  "vehicle_license_province": "กรุงเทพมหานคร",
  "vehicle_model": "PAJERO SPORT",
  "vehicle_model_year": "2017",
  "vehicle_piston": "4",
  "vehicle_register_date": "26 มกราคม 2561",
  "vehicle_seat_number": "7",
  "vehicle_shaft": "2 เพลา 4 ล้อ ยาง 4 เส้น",
  "vehicle_shaft_weight": "",
  "vehicle_short_type": "51",
  "vehicle_type": "รถยนต์นั่งส่วนบุคคลไม่เกิน 7 คน",
}
`;

exports[`process 6.json return correctly 1`] = `
Object {
  "confidence": Object {
    "vehicle_body_type": 0.82,
    "vehicle_brand": 0.91,
    "vehicle_car_weight": 0,
    "vehicle_chassis_location": 0.25,
    "vehicle_chassis_number": 0.81,
    "vehicle_color": 0,
    "vehicle_displacement": 0,
    "vehicle_engine_brand": 0.99,
    "vehicle_engine_location": 0.46,
    "vehicle_engine_number": 0.8,
    "vehicle_fuel_type": 0.68,
    "vehicle_gas_number": 0,
    "vehicle_gvw": 0,
    "vehicle_hp": 0,
    "vehicle_license_number": 0,
    "vehicle_license_province": 0.92,
    "vehicle_model": 0.95,
    "vehicle_model_year": 0.6,
    "vehicle_piston": 0,
    "vehicle_register_date": 0.71,
    "vehicle_seat_number": 0,
    "vehicle_shaft": 0.84,
    "vehicle_shaft_weight": 0,
    "vehicle_short_type": 0,
    "vehicle_type": 0.87,
  },
  "vehicle_body_type": "รถจักรยานยนต์",
  "vehicle_brand": "HONDA",
  "vehicle_car_weight": "",
  "vehicle_chassis_location": "โตเกง",
  "vehicle_chassis_number": "HLHJF350XK5629846",
  "vehicle_color": "",
  "vehicle_displacement": "",
  "vehicle_engine_brand": "HONDA",
  "vehicle_engine_location": "เข้ายเลรอง",
  "vehicle_engine_number": "JT35E-022546",
  "vehicle_fuel_type": "เบนซิน",
  "vehicle_gas_number": "",
  "vehicle_gvw": "",
  "vehicle_hp": "",
  "vehicle_license_number": "",
  "vehicle_license_province": "กรุงเทพมหานคร",
  "vehicle_model": "CLICK 125 1",
  "vehicle_model_year": "2013",
  "vehicle_piston": "",
  "vehicle_register_date": "11 มิถุนายน 2552",
  "vehicle_seat_number": "",
  "vehicle_shaft": "- เพลา 2 ล้อ ยาง 2 เส้น2",
  "vehicle_shaft_weight": "",
  "vehicle_short_type": "",
  "vehicle_type": "รถจักรยานยนต์",
}
`;

exports[`process 7.json return correctly 1`] = `
Object {
  "confidence": Object {
    "vehicle_body_type": 0.99,
    "vehicle_brand": 0.95,
    "vehicle_car_weight": 0,
    "vehicle_chassis_location": 0.38,
    "vehicle_chassis_number": 0.87,
    "vehicle_color": 0.26,
    "vehicle_displacement": 0.96,
    "vehicle_engine_brand": 0.99,
    "vehicle_engine_location": 0.88,
    "vehicle_engine_number": 0.72,
    "vehicle_fuel_type": 0.79,
    "vehicle_gas_number": 0,
    "vehicle_gvw": 0,
    "vehicle_hp": 0,
    "vehicle_license_number": 0.83,
    "vehicle_license_province": 0.96,
    "vehicle_model": 0.96,
    "vehicle_model_year": 1,
    "vehicle_piston": 0.95,
    "vehicle_register_date": 0.97,
    "vehicle_seat_number": 0.99,
    "vehicle_shaft": 0.98,
    "vehicle_shaft_weight": 0,
    "vehicle_short_type": 0.95,
    "vehicle_type": 0.99,
  },
  "vehicle_body_type": "รถจักรยานยนต์",
  "vehicle_brand": "YAMAHA",
  "vehicle_car_weight": "",
  "vehicle_chassis_location": "ใต้ที่นั่ง",
  "vehicle_chassis_number": "MH3RG472111001906",
  "vehicle_color": "การ",
  "vehicle_displacement": "155",
  "vehicle_engine_brand": "YAMAHA",
  "vehicle_engine_location": "ช้ายเครื่อง",
  "vehicle_engine_number": "G3JEE-0012705",
  "vehicle_fuel_type": "เบนซิน",
  "vehicle_gas_number": "",
  "vehicle_gvw": "",
  "vehicle_hp": "",
  "vehicle_license_number": "5กฮ 908",
  "vehicle_license_province": "กรุงเทพมหานคร",
  "vehicle_model": "YZF-R15",
  "vehicle_model_year": "2017",
  "vehicle_piston": "1",
  "vehicle_register_date": "24 กรกฎาคม 2560",
  "vehicle_seat_number": "2",
  "vehicle_shaft": "2 เพลา 2 ล้อ ยาง 2 เส้น",
  "vehicle_shaft_weight": "",
  "vehicle_short_type": "12",
  "vehicle_type": "รถจักรยานยนต์",
}
`;

exports[`process 8.json return correctly 1`] = `
Object {
  "confidence": Object {
    "vehicle_body_type": 0.7,
    "vehicle_brand": 0.97,
    "vehicle_car_weight": 0,
    "vehicle_chassis_location": 0.82,
    "vehicle_chassis_number": 0.78,
    "vehicle_color": 0.71,
    "vehicle_displacement": 0.95,
    "vehicle_engine_brand": 0.99,
    "vehicle_engine_location": 0.76,
    "vehicle_engine_number": 0.71,
    "vehicle_fuel_type": 0.5,
    "vehicle_gas_number": 0,
    "vehicle_gvw": 0,
    "vehicle_hp": 0,
    "vehicle_license_number": 0.78,
    "vehicle_license_province": 0.81,
    "vehicle_model": 0.36,
    "vehicle_model_year": 0.99,
    "vehicle_piston": 0.81,
    "vehicle_register_date": 0.97,
    "vehicle_seat_number": 0.65,
    "vehicle_shaft": 0.93,
    "vehicle_shaft_weight": 0,
    "vehicle_short_type": 0.98,
    "vehicle_type": 0.98,
  },
  "vehicle_body_type": "รถจักรยานยนต์",
  "vehicle_brand": "YAMAHA",
  "vehicle_car_weight": "",
  "vehicle_chassis_location": ",ใต้ที่นั่ง",
  "vehicle_chassis_number": "HH3RG472111001906",
  "vehicle_color": "ดา",
  "vehicle_displacement": "155",
  "vehicle_engine_brand": "YAMAHA",
  "vehicle_engine_location": "ซ้ายเครื่อง",
  "vehicle_engine_number": "G3JSE-0012705",
  "vehicle_fuel_type": "เขนชิ้น",
  "vehicle_gas_number": "",
  "vehicle_gvw": "",
  "vehicle_hp": "",
  "vehicle_license_number": "5กช 808",
  "vehicle_license_province": "กรุงเทพมหานคร",
  "vehicle_model": "YP 5",
  "vehicle_model_year": "2017",
  "vehicle_piston": "1",
  "vehicle_register_date": "24 กรกฎาคม 2560",
  "vehicle_seat_number": "2 คน",
  "vehicle_shaft": "2 เพลา 2 ล้อ ยาง 2 เส้น",
  "vehicle_shaft_weight": "",
  "vehicle_short_type": "12",
  "vehicle_type": "รถจักรยานยนต์",
}
`;

exports[`process 12.json return correctly 1`] = `
Object {
  "confidence": Object {
    "vehicle_body_type": 0.98,
    "vehicle_brand": 0.93,
    "vehicle_car_weight": 0,
    "vehicle_chassis_location": 0.79,
    "vehicle_chassis_number": 0.92,
    "vehicle_color": 0.96,
    "vehicle_displacement": 0.98,
    "vehicle_engine_brand": 0.99,
    "vehicle_engine_location": 0.8,
    "vehicle_engine_number": 0.91,
    "vehicle_fuel_type": 0.57,
    "vehicle_gas_number": 0,
    "vehicle_gvw": 0,
    "vehicle_hp": 0,
    "vehicle_license_number": 0.82,
    "vehicle_license_province": 0.95,
    "vehicle_model": 0.99,
    "vehicle_model_year": 0.91,
    "vehicle_piston": 0.97,
    "vehicle_register_date": 0.98,
    "vehicle_seat_number": 0.96,
    "vehicle_shaft": 0.99,
    "vehicle_shaft_weight": 0,
    "vehicle_short_type": 0.99,
    "vehicle_type": 0.98,
  },
  "vehicle_body_type": "รถจักรยานยนต์",
  "vehicle_brand": "YAMAHA",
  "vehicle_car_weight": "",
  "vehicle_chassis_location": "ใต้ที่นั่ง",
  "vehicle_chassis_number": "MH3RG472111001906",
  "vehicle_color": "ดำ",
  "vehicle_displacement": "155",
  "vehicle_engine_brand": "YAMAHA",
  "vehicle_engine_location": "ซ้ายเครื่อง",
  "vehicle_engine_number": "G3J6E-0012705",
  "vehicle_fuel_type": "เขนชิน",
  "vehicle_gas_number": "",
  "vehicle_gvw": "",
  "vehicle_hp": "",
  "vehicle_license_number": "5กฮ 908",
  "vehicle_license_province": "กรุงเทพมหานคร",
  "vehicle_model": "YZF-R15",
  "vehicle_model_year": "2017",
  "vehicle_piston": "1",
  "vehicle_register_date": "24 กรกฎาคม 2560",
  "vehicle_seat_number": "2",
  "vehicle_shaft": "2 เพลา 2 ล้อ ยาง 2 เส้น",
  "vehicle_shaft_weight": "",
  "vehicle_short_type": "12",
  "vehicle_type": "รถจักรยานยนต์",
}
`;

exports[`process 13.json return correctly 1`] = `
Object {
  "confidence": Object {
    "vehicle_body_type": 0.83,
    "vehicle_brand": 0.99,
    "vehicle_car_weight": 0,
    "vehicle_chassis_location": 0.83,
    "vehicle_chassis_number": 0.56,
    "vehicle_color": 0.99,
    "vehicle_displacement": 0.92,
    "vehicle_engine_brand": 0.89,
    "vehicle_engine_location": 0.85,
    "vehicle_engine_number": 0.85,
    "vehicle_fuel_type": 0.83,
    "vehicle_gas_number": 0,
    "vehicle_gvw": 0,
    "vehicle_hp": 0,
    "vehicle_license_number": 0.93,
    "vehicle_license_province": 0.99,
    "vehicle_model": 0.96,
    "vehicle_model_year": 0.99,
    "vehicle_piston": 0.98,
    "vehicle_register_date": 0.89,
    "vehicle_seat_number": 0.93,
    "vehicle_shaft": 0.96,
    "vehicle_shaft_weight": 0,
    "vehicle_short_type": 0.92,
    "vehicle_type": 0.96,
  },
  "vehicle_body_type": "รถจักรยานยนต์",
  "vehicle_brand": "HONDA",
  "vehicle_car_weight": "",
  "vehicle_chassis_location": "ใต้ที่นั่ง",
  "vehicle_chassis_number": "MLHJF350XK5629846 .",
  "vehicle_color": "ดำ",
  "vehicle_displacement": "124.88",
  "vehicle_engine_brand": "HONDA",
  "vehicle_engine_location": "ช้ายเครื่อง",
  "vehicle_engine_number": "JF35E-0229846",
  "vehicle_fuel_type": "เขนชิน",
  "vehicle_gas_number": "",
  "vehicle_gvw": "",
  "vehicle_hp": "",
  "vehicle_license_number": "กร 7876",
  "vehicle_license_province": "กรุงเทพมหานคร",
  "vehicle_model": "CLICK 125 1",
  "vehicle_model_year": "2019",
  "vehicle_piston": "1",
  "vehicle_register_date": "11 มิถุนายน 2552",
  "vehicle_seat_number": "2",
  "vehicle_shaft": "- เพลา 2 ล้อ ยาง 2 เส้น",
  "vehicle_shaft_weight": "",
  "vehicle_short_type": "12",
  "vehicle_type": "รถจักรยานยนต์",
}
`;

exports[`process 14.json return correctly 1`] = `
Object {
  "confidence": Object {
    "vehicle_body_type": 0.93,
    "vehicle_brand": 0.97,
    "vehicle_car_weight": 0,
    "vehicle_chassis_location": 0.64,
    "vehicle_chassis_number": 0.94,
    "vehicle_color": 0.95,
    "vehicle_displacement": 0.94,
    "vehicle_engine_brand": 0.99,
    "vehicle_engine_location": 0.82,
    "vehicle_engine_number": 0.91,
    "vehicle_fuel_type": 0.73,
    "vehicle_gas_number": 0,
    "vehicle_gvw": 0,
    "vehicle_hp": 0,
    "vehicle_license_number": 0.65,
    "vehicle_license_province": 0.86,
    "vehicle_model": 0.92,
    "vehicle_model_year": 0.98,
    "vehicle_piston": 0.92,
    "vehicle_register_date": 0.83,
    "vehicle_seat_number": 0.99,
    "vehicle_shaft": 0.99,
    "vehicle_shaft_weight": 0,
    "vehicle_short_type": 0.87,
    "vehicle_type": 0.99,
  },
  "vehicle_body_type": "รถจักรยานยนต์",
  "vehicle_brand": "HONDA",
  "vehicle_car_weight": "",
  "vehicle_chassis_location": "เสียงใต้ที่นั่ง",
  "vehicle_chassis_number": "HLHJF350XK5629846",
  "vehicle_color": "ดำ",
  "vehicle_displacement": "124.88",
  "vehicle_engine_brand": "HONDA",
  "vehicle_engine_location": "ช้ายเครื่อง",
  "vehicle_engine_number": "JF 35E-0229846",
  "vehicle_fuel_type": "เขนซิน",
  "vehicle_gas_number": "",
  "vehicle_gvw": "",
  "vehicle_hp": "",
  "vehicle_license_number": "9กร 78763กร 7876",
  "vehicle_license_province": "ในหานครกรุงเทพมหานคร",
  "vehicle_model": "CLICK 125 I",
  "vehicle_model_year": "2019",
  "vehicle_piston": "1",
  "vehicle_register_date": "ายน 2511 มิถุนายน 2562",
  "vehicle_seat_number": "2",
  "vehicle_shaft": "- เพลา 2 ล้อ ยาง 2 เส้น",
  "vehicle_shaft_weight": "",
  "vehicle_short_type": "12",
  "vehicle_type": "รถจักรยานยนต์",
}
`;
